### PR TITLE
DM-37908: Data Management Initial integration of observing blocks to the Scheduler

### PR DESCRIPTION
**Pull request by tribeiro**
A couple notes:

* I have not organized the commits yet, or written down the version history. There's quite a lot of code changes so now the code changes are done I want to get it in review.
* The CI job will fail until the configuration repo is merged, because the Scheduler CI is not setup to checkout the branch. The PR for the configuration is [here](https://github.com/lsst-ts/ts_config_ocs/pull/145) but it is probably best to wait until we are done with this review before merging it so, the CI will be failing in the meantime. All tests passes in my local environment.

### CODE REVIEW

File: python/lsst/ts/scheduler/driver/sequential.py [hunk 1]
```diff
@@ -119,17 +83,41 @@ class SequentialScheduler(Driver):
     The driver reads from an input file of targets provided by the user and
     send one target at a time.
 
+    Parameters
+    ----------
+    models : dict[str, typing.Any]
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/driver/sequential.py:88]: Please enclose attribute types in back ticks.

File: python/lsst/ts/scheduler/config_schema.py [hunk 2]
```diff
@@ -26,623 +26,684 @@
...
+                  altitude_accel:
+                    type: number
+                    description: Dome acceleration in elevation (degrees/s^2).
+                  altitude_decel:
+                    type: number
+                    description: Dome deceleration in elevation (degrees/s^2).
+                  altitude_freerange:
+                    type: number
+                    description: >-
+                      Dome slit free-range in elevation (degrees). Specify how much
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/config_schema.py:141]: `Specify` -> `Specifies`

File: python/lsst/ts/scheduler/config_schema.py [hunk 3]
```diff
@@ -26,623 +26,684 @@
...
+                  azimuth_accel:
+                    type: number
+                    description: Dome acceleration in azimuth (degrees/s^2).
+                  azimuth_decel:
+                    type: number
+                    description: Dome deceleration in azimuth (degrees/s^2).
+                  azimuth_freerange:
+                    type: number
+                    description: >-
+                      Dome slit free-range in azimuth (degrees). Specify how much the
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/config_schema.py:156]: `Specify` -> `Specifies`

File: python/lsst/ts/scheduler/model.py [hunk 4]
```diff
@@ -57,49 +56,76 @@ class Model:
     instantiates the model and offload the operations to it, which allows for
     better separation of concerns and cleaner code on the CSC part.
 
+    Parameters
+    ----------
+    log : logging.Logger
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/model.py:61]: Please enclose attribute types in back ticks, here and further below.

File: python/lsst/ts/scheduler/model.py [hunk 5]
```diff
@@ -190,6 +218,29 @@ async def configure_telemetry_streams(self, config: dict[str, typing.Any]) -> No
...
 
+    async def load_observing_blocks(self, path: str) -> None:
+        """Load observing blocks from the provided path.
+
+        The method will glob the provided path for all json files and load them
+        as observing blocks.
+
+        Parameters
+        ----------
+        path : str
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/model.py:229]: Please enclose attribute types in back ticks.

File: python/lsst/ts/scheduler/model.py [hunk 6]
```diff
@@ -315,51 +366,21 @@ async def configure_driver_cold(self, config: typing.Any) -> SurveyTopology:
         return survey_topology
 
     def load_driver(self, driver_type: str) -> None:
-        """Utility method to load a driver from a driver type.
+        """Utility method to load a driver from a driver type name.
 
         Parameters
         ----------
         driver_type : str
-            A driver module "import string", e.g.
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/model.py:323]: Please enclose attribute types in back ticks.

File: python/lsst/ts/scheduler/utils/fbs_utils.py [hunk 7]
```diff
@@ -164,3 +163,26 @@ def opsim2obs(self, filename: str) -> np.ndarray:
                 final_result[key] = df[key].values
 
         return final_result
+
+    def opsim2df(self, filename: str) -> pd.DataFrame:
+        """Read an opsim database and return a pandas data frame.
+
+        Parameters
+        ----------
+        filename : str
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/utils/fbs_utils.py:172]: Please enclose attribute types in back ticks.

File: tests/test_driver_target.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
+# This file is part of ts_scheduler
+#
+# Developed for the LSST Telescope and Site Systems.
```
Comment (wvreeven) [inline tests/test_driver_target.py:3]: `Vera Rubin Observatory` please.

File: tests/test_driver_target.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
```
Comment (wvreeven) [inline tests/test_driver_target.py:19]: This header is truncated. Please include a complete header.

**Review by wvreeven — APPROVED** [review]
Looks good to me in principle. I left several minor comments for your consideration.

File: python/lsst/ts/scheduler/driver/driver.py [hunk 10]
```diff
@@ -135,13 +138,14 @@ def __init__(
         self.raw_telemetry = raw_telemetry
         self.targetid = 0
 
-        self.default_observing_script_name = None
-        self.default_observing_script_is_standard = None
+        self.observing_blocks = observing_blocks
+        # self.default_observing_script_name = None
```
[the remark this answers is not in the document]
Comment (tribeiro) [reply python/lsst/ts/scheduler/driver/driver.py:142]: Good catch! thanks! I just forgot it here..

File: python/lsst/ts/scheduler/driver/driver_target.py [hunk 11]
```diff
@@ -100,51 +95,145 @@ def __init__(
...
-            return ""
+        return self._sal_indices
 
-        script_config = {
+    def add_sal_index(self, sal_index: int) -> None:
+        """Add SAL index.
+
+        Parameters
+        ----------
+        sal_index : int
```
[the remark this answers is not in the document]
Comment (tribeiro) [reply python/lsst/ts/scheduler/driver/driver_target.py:116]: is this really necessary? In the numpy docs [style guide](https://numpydoc.readthedocs.io/en/latest/format.html) they don't use back ticks..

File: python/lsst/ts/scheduler/driver/feature_scheduler_target.py [hunk 12]
```diff
@@ -18,53 +18,36 @@
...
+import numpy as np
 
-import yaml
-from astropy import units
-from astropy.coordinates import Angle
+from lsst.ts import observing
 
 from .driver_target import DriverTarget
 
+HOURANGLE_2_DEG = 180.0 / 12.0
```
[the remark this answers is not in the document]
Comment (tribeiro) [reply python/lsst/ts/scheduler/driver/feature_scheduler_target.py:27]: This is not needed so I removed it. Thanks for pointing it out!

File: python/lsst/ts/scheduler/driver/driver_target.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tribeiro) [reply python/lsst/ts/scheduler/driver/driver_target.py:116]: Ok, found Tre rational behind it: https://developer.lsst.io/python/numpydoc.html#marking-up-parameter-names

I'll add back ticks..

**Pull request merged**